File: badart.py
#!/usr/bin/python3

import argparse
import asyncio
import json
import os
import time
import unicodedata

import http.client
import tornado.web

import scrum


class Image:
  def __init__(self, url, width):
    self.url = url
    self.width = width
    self.last = False

class Painting:
  BY_NAME = {}

  ORDERED_IMAGES = []

  @classmethod
  def set_options(cls, options):
    cls.options = options

  def __init__(self, title, answers, image_dir):
    self.title = title  # with "Bad"

    self.answers = []
    for a in answers:
      self.answers.append(a)
      if a.startswith("THE"):
        self.answers.append("THEBAD" + a[3:])
      else:
        self.answers.append("BAD" + a)

    self.images = []
    self.solved = False

    images = [i for i in self.options.assets.keys()
              if i.startswith(image_dir+"/") and i.endswith(".png")]
    images.sort()

    for im in images:
      self.images.append(Image(self.options.assets[im], 800))
      self.ORDERED_IMAGES.append(self.images[-1].url)

    self.images[-1].last = True

class Message:
  def __init__(self, serial, message):
    self.serial = serial
    self.timestamp = time.time()
    self.message = message

class GameState:
  BY_TEAM = {}

  FRAME_SECS = 1.5
  LAST_SECS = 4.0
  INITIAL_OPENING_TIME = 600  # 10 minutes
  CLOSURE_TIME = 1800  # 30 minutes

  @classmethod
  def set_globals(cls, paintings, preload, options):
    cls.paintings = paintings
    cls.preload = preload
    cls.options = options

  @classmethod
  def get_for_team(cls, team):
    if team not in cls.BY_TEAM:
      cls.BY_TEAM[team] = cls(team)
    return cls.BY_TEAM[team]

  def __init__(self, team):
    self.team = team
    self.sessions = set()
    self.running = False
    self.cond = asyncio.Condition()
    self.current_painting = None
    self.open_requested = False

  async def on_wait(self, session):
    async with self.cond:
      if session not in self.sessions:
        self.sessions.add(session)
        self.cond.notify_all()

  async def run_game(self):
    while not self.open_requested:
      count = len(self.sessions)
      text = (f"{count} player{' is' if count == 1 else 's are'} currently waiting.<br>"
              f"You can enter the gallery when there are {self.options.min_players}.")

      if len(self.sessions) < self.options.min_players:
        msg = {"method": "show_message", "text": text}
      else:
        msg = {"method": "prompt_open", "text": text}

      await self.team.send_messages([msg], sticky=1)
      async with self.cond:
        await self.cond.wait()

    now = time.time()
    close_time = now + self.INITIAL_OPENING_TIME
    reopen_time = close_time + self.CLOSURE_TIME
    state = "initial_open"

    while True:
      just_solved = False
      for p in self.paintings:
        self.current_painting = p
        while True:
          next_painting = True
          if p.solved:
            to_show = p.images[-1:]
          else:
            to_show = p.images

          for i in to_show:

            if state == "initial_open" and time.time() > close_time:
              state = "closed"
              d = {"method": "show_message",
                   "text": "The gallery is now closed.<br>It will reopen in 30 minutes.",
                   "end_time": reopen_time,
                   "countdown_text": "The gallery will reopen in"}
              await self.team.send_messages([d], sticky=1)
              await asyncio.sleep(reopen_time - time.time())
              close_time = None
              state = "reopened"

            d = {"method": "show_image",
                 "image": i.url,
                 "preload": self.preload[i.url],
                 "width": str(i.width) + "px",
            }
            if close_time:
              d["end_time"] = close_time
            if p.solved:
              d["title"] = p.title

            if just_solved:
              to_send = [{"method": "play_audio",
                          "url": self.options.assets["tada.wav"]}, d]
              just_solved = False
            else:
              to_send = [d]
            await self.team.send_messages(to_send, sticky=1)

            delay = self.LAST_SECS if i.last else self.FRAME_SECS
            async with self.cond:
              was_solved = p.solved
              try:
                await asyncio.wait_for(self.cond.wait(), delay)
              except asyncio.TimeoutError:
                pass    # next image in current painting
              else:
                if p.solved and not was_solved:
                  # restart current painting to show solved state
                  just_solved = True
                  next_painting = False
                  break

          if next_painting: break




  async def send_chat(self, text):
    d = {"method": "add_chat", "text": text}
    await self.team.send_messages([d])

  async def try_answer(self, answer):
    async with self.cond:
      if (not self.current_painting.solved and
          answer in self.current_painting.answers):
        self.current_painting.solved = True
        self.cond.notify_all()

  async def request_open(self):
    async with self.cond:
      self.open_requested = True
      self.cond.notify_all()


class BadArtApp(scrum.ScrumApp):
  async def on_wait(self, team, session):
    gs = GameState.get_for_team(team)

    if not gs.running:
      gs.running = True
      self.add_callback(gs.run_game)

    await gs.on_wait(session)


class SubmitHandler(tornado.web.RequestHandler):
  def prepare(self):
    self.args = json.loads(self.request.body)

  @staticmethod
  def canonicalize_answer(text):
    text = unicodedata.normalize("NFD", text.upper())
    out = []
    for k in text:
      cat = unicodedata.category(k)
      # Letters and "other symbols".
      if cat == "So" or cat == "Nd" or cat[0] == "L":
        out.append(k)
    return "".join(out)

  async def post(self):
    scrum_app = self.application.settings["scrum_app"]
    team, session = await scrum_app.check_cookie(self)
    gs = GameState.get_for_team(team)

    submission = self.args["answer"]
    answer = self.canonicalize_answer(submission)
    who = self.args["who"].strip()
    if not who: who = "anonymous"
    print(f"{team}: {who} submitted {answer}")

    await gs.send_chat(f"{who} guessed \"{submission}\"")
    await gs.try_answer(answer)

    self.set_status(http.client.NO_CONTENT.value)


class OpenHandler(tornado.web.RequestHandler):
  async def get(self):
    scrum_app = self.application.settings["scrum_app"]
    team, session = await scrum_app.check_cookie(self)
    gs = GameState.get_for_team(team)
    await gs.request_open()
    self.set_status(http.client.NO_CONTENT.value)


class DebugHandler(tornado.web.RequestHandler):
  def get(self, fn):
    if fn.endswith(".css"):
      self.set_header("Content-Type", "text/css")
    elif fn.endswith(".js"):
      self.set_header("Content-Type", "application/javascript")
    with open(fn) as f:
      self.write(f.read())


def make_app(options):
  Painting.set_options(options)

  paintings = [
    Painting("Bad Flag (Moratorium)", {"FLAGMORATORIUM"}, "01"),
    Painting("Bad Irises", {"IRISES"}, "02"),
    Painting("Bad Nighthawks", {"NIGHTHAWKS"}, "03"),
    Painting("Bad Guernica", {"GUERNICA"}, "04"),
    Painting("The Bad Emperor Napoleon in His Study at the Tuileries",
             {"THEEMPERORNAPOLEONINHISSTUDYATTHETUILERIES",
              "EMPERORNAPOLEONINHISSTUDYATTHETUILERIES"},
             "05"),
    Painting("The Bad Raft of the Medusa",
             {"THERAFTOFTHEMEDUSA",
              "RAFTOFTHEMEDUSA",
              "RAFTOFMEDUSA"},
             "06"),
    Painting("The Bad Persistence of Memory",
             {"THEPERSISTENCEOFMEMORY",
              "PERSISTENCEOFMEMORY"},
             "07"),
    Painting("The Bad Arnolfini Portrait",
             {"ARNOLFINIPORTRAIT",
              "THEARNOLFINIPORTRAIT",
              "ARNOLFINIWEDDING",
              "THEARNOLFINIWEDDING",
              "ARNOLFINIMARRIAGE",
              "THEARNOLFINIMARRIAGE"},
             "08"),
    Painting("Bad Impression, Sunrise",
             {"IMPRESSIONSUNRISE"},
             "09"),
    Painting("The Bad Night Watch",
             {"THENIGHTWATCH",
              "NIGHTWATCH"},
             "10"),
    Painting("The Bad Third of May 1808",
             {"THETHIRDOFMAY1808",
              "THIRDOFMAY1808"},
             "11"),
    Painting("The Bad Son of Man",
             {"THESONOFMAN", "SONOFMAN"},
             "12"),
  ]

  preload = {}
  oi = Painting.ORDERED_IMAGES
  for i, u in enumerate(oi):
    preload[u] = oi[(i+1)%len(oi)]

  GameState.set_globals(paintings, preload, options)

  handlers = [
    (r"/artsubmit", SubmitHandler),
    (r"/artopen", OpenHandler),
    (r"/artdebug/(\S+)", DebugHandler),
  ]

  return handlers


def main():
  parser = argparse.ArgumentParser(description="Run the bad art puzzle.")
  parser.add_argument("--assets_json", default=None,
                      help="JSON file for image assets")
  parser.add_argument("-c", "--cookie_secret",
                      default="snellen2020",
                      help="Secret used to create session cookies.")
  parser.add_argument("--socket_path", default="/tmp/badart",
                      help="Socket for requests from frontend.")
  parser.add_argument("--wait_url", default="artwait",
                      help="Path for wait requests from frontend.")
  parser.add_argument("--main_server_port", type=int, default=2020,
                      help="Port to use for requests to main server.")
  parser.add_argument("--min_players", type=int, default=1,
                      help="Number of players needed to start game.")

  options = parser.parse_args()

  assert options.assets_json
  with open(options.assets_json) as f:
    options.assets = json.load(f)

  app = BadArtApp(options, make_app(options))
  app.start()


if __name__ == "__main__":
  main()

File: badart.js
goog.require('goog.dom');
goog.require('goog.dom.classlist');
goog.require('goog.dom.TagName');
goog.require('goog.events');
goog.require('goog.events.KeyCodes');
goog.require('goog.net.XhrIo');
goog.require("goog.json.Serializer");

class BadArtCountdown {
    constructor(msg, end_time) {
	this.msg = msg;
	this.end_time = end_time;
	this.timer = setInterval(goog.bind(this.update_timer, this), 1000);
    }

    update_timer() {
	var now = (new Date()).getTime() / 1000.0;
	var s = this.end_time - now;
	if (s < 0) s = 0;
	var min = Math.trunc(s/60);
	var sec = Math.trunc(s%60);
	var text = this.msg + " " + min + ":" + (""+sec).padStart(2, "0") + ".";
	badart.countdown_span.innerHTML = text;
    }

    finish() {
	clearInterval(this.timer);
	badart.countdown_span.innerHTML = "";
    }
}


class BadArtDispatcher {
    constructor() {
	this.methods = {
	    "show_message": this.show_message,
	    "prompt_open": this.prompt_open,
	    "show_image": this.show_image,
	    "play_audio": this.play_audio,
	    "add_chat": this.add_chat,
            "players": this.players,
	}
    }

    /** @param{Message} msg */
    dispatch(msg) {
	this.methods[msg.method](msg);
    }

    /** @param{Message} msg */
    show_message(msg) {
	badart.gallery.style.display = "none";
	badart.entry.style.display = "none";
	badart.message.style.display = "block";
	badart.message.innerHTML = msg.text;

	if (msg.end_time) {
	    if (badart.countdown) {
		badart.countdown.finish();
	    }
	    badart.countdown = new BadArtCountdown(msg.countdown_text, msg.end_time);
	}
    }

    /** @param{Message} msg */
    prompt_open(msg) {
	badart.open.style.display = "inline";
	badart.gallery.style.display = "none";
	badart.entry.style.display = "none";
	badart.message.style.display = "block";
	badart.message.innerHTML = msg.text;
    }

    /** @param{Message} msg */
    players(msg) {
        var el = goog.dom.getElement("players");
        el.innerHTML = "<b>Players:</b> " + msg.players;
    }

    /** @param{Message} msg */
    show_image(msg) {
	if (msg.end_time) {
	    if (badart.countdown == null) {
		badart.countdown = new BadArtCountdown("The gallery will close in", msg.end_time);
	    }
	} else {
	    if (badart.countdown) {
		badart.countdown.finish();
	    }
	}
	badart.gallery.style.display = "block";
	badart.message.style.display = "none";
	badart.open.style.display = "none";
	if (msg.title) {
	    goog.dom.classlist.add(badart.art, "framed");
	    badart.caption.style.display = "flex";
	    badart.entry.style.display = "none";
	    badart.title.innerHTML = msg.title;
	    badart.text.value = "";
	} else {
	    goog.dom.classlist.remove(badart.art, "framed");
	    badart.caption.style.display = "none";
	    badart.entry.style.display = "flex";
	    if (document.activeElement.tagName != "INPUT") badart.text.focus();
	}

	badart.art.src = msg.image;
	badart.preload.src = msg.preload;
    }

    /** @param{Message} msg */
    play_audio(msg) {
	if (msg.url) {
	    var audio = new Audio(msg.url);
	    audio.play();
	}
    }

    /** @param{Message} msg */
    add_chat(msg) {
	var curr = goog.dom.getChildren(badart.chat);
	if (curr.length > 3) {
	    goog.dom.removeNode(curr[0]);
	}
	var el = goog.dom.createDom("P");
        el.innerHTML = msg.text;
	badart.chat.appendChild(el);
    }
}

function badart_submit(e) {
    var answer = badart.text.value;
    if (answer == "") return;
    badart.text.value = "";
    var username = badart.who.value;
    localStorage.setItem("name", username);
    var msg = badart.serializer.serialize({"answer": answer, "who": username});
    goog.net.XhrIo.send("/artsubmit", Common_expect_204, "POST", msg);
    e.preventDefault();
}


function badart_onkeydown(e) {
    if (e.keyCode == goog.events.KeyCodes.ENTER) {
	badart_submit(e);
    }
}

function badart_send_name() {
    var name = badart.who.value;
    if (name != badart.sent_name) {
        badart.sent_name = name;
        var msg = badart.serializer.serialize({"who": name});
        goog.net.XhrIo.send("/artname", Common_expect_204, "POST", msg);
    }
}

var badart = {
    body: null,
    gallery: null,
    open: null,
    art: null,
    caption: null,
    entry: null,
    title: null,
    waiter: null,
    text: null,
    who: null,
    chat: null,
    message: null,
    preload: null,
    countdown_span: null,
    countdown: null,
    serializer: null,
    sent_name: null,
}

puzzle_init = function() {
    badart.serializer = new goog.json.Serializer();

    badart.body = goog.dom.getElement("puzz");
    badart.gallery = goog.dom.getElement("gallery");
    badart.open = goog.dom.getElement("open");
    badart.art = goog.dom.getElement("art");
    badart.caption = goog.dom.getElement("caption");
    badart.entry = goog.dom.getElement("entry");
    badart.text = goog.dom.getElement("text");
    badart.title = goog.dom.getElement("title");
    badart.who = goog.dom.getElement("who");
    badart.who.value = localStorage.getItem("name");
    badart.chat = goog.dom.getElement("chat");
    badart.message = goog.dom.getElement("message");
    badart.countdown_span = goog.dom.getElement("countdown");
    badart.preload = goog.dom.getElement("preload");

    goog.events.listen(goog.dom.getElement("text"),
		       goog.events.EventType.KEYDOWN,
		       badart_onkeydown);
    goog.events.listen(goog.dom.getElement("artsubmit"),
		       goog.events.EventType.CLICK,
		       badart_submit);
    goog.events.listen(badart.open,
		       goog.events.EventType.CLICK,
		       goog.bind(goog.net.XhrIo.send, null, "/artopen"));

    badart.waiter = new Common_Waiter(new BadArtDispatcher(), "/artwait", 0, null, null);
    badart.waiter.start();

    setInterval(badart_send_name, 1000);
}

